src/tracing/perfetto/perfetto_tracing.cpp

```diff
@@ -0,0 +1,242 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <tracing/perfetto/perfetto_tracing.h>
+
+#include <lib/address_resolve/TracingStructs.h>
+#include <lib/support/ErrorStr.h>
+#include <lib/support/StringBuilder.h>
+#include <transport/TracingStructs.h>
+
+#include <perfetto.h>
+#include <matter/tracing/macros_impl.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+
+namespace chip {
+namespace Tracing {
+namespace Perfetto {
+
+PerfettoBackend & PerfettoBackend::Init(const char * output_name)
+{
+    perfetto::TracingInitArgs args;
+
+    if ((output_name != nullptr) && (output_name[0] != '\0'))
+    {
+        OpenTracingFile(output_name);
+        args.backends = perfetto::kInProcessBackend;
+    }
+    else
+    {
+        args.backends = perfetto::kSystemBackend;
+    }
+
+    perfetto::Tracing::Initialize(args);
+    perfetto::TrackEvent::Register();
+
+    return *this;
+}
+
+PerfettoBackend::~PerfettoBackend()
+{
+    CloseTracingFile();
+}
+
+void PerfettoBackend::OpenTracingFile(const char * name)
+{
+    VerifyOrDie(name != nullptr);
+    CloseTracingFile();
+
+    mTraceFileId = open(name, O_RDWR | O_CREAT | O_TRUNC, 0640);
+    if (mTraceFileId < 0)
+    {
+        ChipLogError(Automation, "Failed to open logging file '%s': %s", name, strerror(errno));
+        mTraceFileId = kInvalidFileId;
+    }
+}
+
+void PerfettoBackend::CloseTracingFile()
+{
+    if (mTraceFileId != kInvalidFileId)
+    {
+        close(mTraceFileId);
+        mTraceFileId = kInvalidFileId;
+    }
+}
+
+void PerfettoBackend::Open()
+{
+    if (mTraceFileId == kInvalidFileId)
+    {
+        return;
+    }
+
+    perfetto::TraceConfig cfg;
+    cfg.add_buffers()->set_size_kb(1024);
+
+    auto * ds_cfg = cfg.add_data_sources()->mutable_config();
+    ds_cfg->set_name("track_event"); // corresponds to TRACE_EVENT
+
+    mTracingSession = perfetto::Tracing::NewTrace();
+    mTracingSession->Setup(cfg, mTraceFileId);
+    mTracingSession->StartBlocking();
+}
+
+void PerfettoBackend::Close()
+{
+    if (!mTracingSession)
+    {
+        return;
+    }
+
+    perfetto::TrackEvent::Flush();
+
+    mTracingSession->FlushBlocking();
+    mTracingSession->StopBlocking();
+    mTracingSession.reset();
+    CloseTracingFile();
+}
+
+void PerfettoBackend::TraceBegin(const char * label, const char * group)
+{
+    // We generally expect macros to be used using
+    //
+    // matter_trace_config = "${chip_root}/src/tracing/perfetto:tracing"
+    //
+    // However perfetto can also do non-const string tracing (just likely
+    // slower)
+    TRACE_EVENT_BEGIN("Matter", perfetto::StaticString(label), "class_name", perfetto::StaticString(group));
```

Comment:
This is inlined currently by macros_impl.h

This is just a version that also allows trace events when trace events are multiplexed. I can also completely remove it if you think that is better.

Reply:
This is confusing and I would not suggest adding a perfetto backend that does not work properly (the argument being that outlining the tracepoints into an encapsulation of the SDK makes it work subtly worse).

If we want to multiplex, wouldn't we just do

```
#define MATTER_TRACE_BEGIN(category, name)
  do {
    #if MATTER_TRACE_PERFETTO
      TRACE_EVENT_BEGIN(category, name)
    #endif
    #if MATTER_TRACE_PIGWEED
      PW_TRACE_START(name, category)
    #endif
    #if  MATTER_TRACE_CUSTOM
      MATTER_TRACE_BEGIN_CUSTOM(category, name)
    #endif
    #if MATTER_DYNAMIC_TRACE_PLUGIN_BUT_DO_WE_ACTUALLY_NEED_IT
      ::chip::TracePluginRegistery->OnTraceBegin(category, name);
    #endif
  } while(0);
```

Because then these options are orthogonal and you can configure a "all in" build if you really want to.

But I don't know if we need the plugins, frankly.

Reply:
In my mind, the perfetto backend does work properly both in multiplexing and in exclusive mode, execept in exclusive mode I likely get line information as well.

At the same time, in the current traces that I see, I see no line or file information at all, so I am unclear what actually breaks.

Could you be more specific? I am looking to decouple logic as much as possible, hence the separate header. What you suggest seems to explicitly couple things by having the top level be aware of every trace type to include in ifdefs. My plan was to avoid that.

Reply:
Removed the TraceBegin/End/Instant from the multiplexed backend, so traces in perfetto will only work if perfetto is the only set backend.